slack-plugin/src/main/kotlin/slack/gradle/artifacts/Publisher.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2024 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package slack.gradle.artifacts
+
+import org.gradle.api.Named
+import org.gradle.api.NamedDomainObjectProvider
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.file.Directory
+import org.gradle.api.file.RegularFile
+import org.gradle.api.provider.Provider
+
+/**
+ * Used for publishing custom artifacts from a subproject to an aggregating project (often the
+ * "root" project). Only for inter-project publishing (e.g., _not_ for publishing to Artifactory).
+ * See also [Resolver].
+ *
+ * Represents a set of tightly coupled [Configuration]s:
+ * * A "dependency scope" configuration ([Resolver.declarable]).
+ * * A "resolvable" configuration ([Resolver.internal]).
+ * * A "consumable" configuration ([external]).
+ *
+ * Dependencies are _declared_ on [Resolver.declarable] in the aggregating project. Custom artifacts
+ * (e.g., not jars), generated by tasks, are published via [publish], which should be used on
+ * dependency (artifact-producing) projects.
+ *
+ * Gradle uses [attributes][Attr] to wire the consumer project's [Resolver.internal] (resolvable)
+ * configuration to the producer project's [external] (consumable) configuration, which is itself
+ * configured via [publish].
+ *
+ * @see <a
+ *   href="https://docs.gradle.org/current/userguide/cross_project_publications.html#sec:variant-aware-sharing">Variant-aware
+ *   sharing of artifacts between projects</a>
+ * @see <a
+ *   href="https://dev.to/autonomousapps/configuration-roles-and-the-blogging-industrial-complex-21mn">Gradle
+ *   configuration roles</a>
```

Comment:
oh that's kind of nice. it always annoyed me that having these on a single line was so long, but this looks readable even in github